# Homework

### Question 4

Download the data (only for podcasts). How many records are there?

In [1]:
import homework as docs

github_data = docs.read_github_data('DataTalksClub', 'datatalksclub.github.io')
parsed_data = docs.parse_data(github_data)


In [2]:
len(github_data)

184

A/ `184 records`

### Question 5

Let's prepare this data. It's already structured, so you can chunk it using paragraphs. Let's do chunk size 30 and overlap 15. How many chunks do you have in the result?

In [152]:
import re

paragraphs = []

for data in github_data:
    text = data.content
    paragraphs.extend(re.split(r"\n\s*\n", text.strip()))

In [153]:
len(paragraphs)

854

In [156]:
def sliding_window(seq, size, step):
    """Create overlapping chunks using sliding window approach."""
    if size <= 0 or step <= 0:
        raise ValueError("size and step must be positive")

    n = len(seq)
    result = []
    for i in range(0, n, step):
        batch = seq[i:i+size]
        result.append(batch)
        if i + size >= n:
            break

    return result

def format_chunk(chunk):
    return {
        'content': ' '.join(chunk)
    }

In [157]:
chunks = []

for chunk in sliding_window(paragraphs, 30, 15):
    processed = format_chunk(chunk)
    chunks.append(processed)


In [158]:
len(chunks)

56

A/ `56 chunks`

### Question 6

Index the data with Index from minsearch. What's the first episode in the results for "how do I make money with AI?"

In [160]:
from minsearch import Index

index = Index(
    text_fields=["content"],
)

index.fit(chunks)

In [166]:
search = index.search("how do I make money with AI?", num_results=5)

In [263]:
print(search[0]['content'][:1200])

---
episode: 8
guests:
- jekaterinakokatjuhha
ids:
  anchor: The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim
  youtube: FRi0SUtxdMw
image: images/podcast/s12e08-journey-of-data-generalist-from-bioinformatics-to-freelancing.jpg
links:
  anchor: https://anchor.fm/datatalksclub/episodes/The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim
  apple: https://podcasts.apple.com/us/podcast/the-journey-of-a-data-generalist-from/id1541710331?i=1000599125044
  spotify: https://open.spotify.com/episode/5fB185hGlGYQmdk0kbIsPv?si=YtnsaYNzTc-fl7emZ2IjEA
  youtube: https://www.youtube.com/watch?v=FRi0SUtxdMw
season: 12
short: 'The Journey of a Data Generalist: From Bioinformatics to Freelancing'
title: 'The Journey of a Data Generalist: From Bioinformatics to Freelancing'
transcript:
- line: This week we'll talk about being a data generalist. We'll discuss going from
    bioinformatics to freelancing. We

A/ `Season 12 - Episode 8 - The Journey of a Data Generalist From Bioinformatics to Freelancing`

### Question 5 Again

In [218]:
from homework import RawRepositoryFile
from typing import List, Dict, Any
import frontmatter

def parse_data(data_raw: List[RawRepositoryFile]) -> List[Dict[str, Any]]:
    data_parsed = []
    for f in data_raw:
        post = frontmatter.loads(f.content)
        data = post.to_dict()
        data['filename'] = f.filename
        data['content'] = data.get('content', '') + ''.join([t.get('line', '') for t in data.get('transcript', [])])
        data_parsed.append(data)

    return data_parsed

In [219]:
parsed_data_2 = parse_data(github_data)

In [221]:
# [{'content': p } for p in paragraphs]
chunks2 = docs.chunk_documents(parsed_data_2, 30, 15)

In [222]:
len(chunks2)

518519

A/ `518519 chunks` 

### Question 6 Again

In [272]:
index2 = Index(
    text_fields=["content"],
)

index2.fit(chunks2)

In [273]:
search2 = index2.search("how do I make money with AI?", num_results=15)

In [274]:
[p['title'] for p in search2]

['The Essentials of Public Speaking for Career in Data Science',
 'Responsible and Explainable AI',
 'Freelancing in Machine Learning',
 'From Open-Source Maintainer to Founder',
 'Lessons Learned from Freelancing and Working in a Start-up',
 'Investing in Open-Source Data Tools',
 'Responsible and Explainable AI',
 'New Roles and Key Skills to Monetize Machine Learning',
 'Freelancing in Machine Learning',
 'Data Science Career Development',
 'From Open-Source Maintainer to Founder',
 'Becoming a Solopreneur in Data',
 'Accelerating the Adoption of AI through Diversity',
 'New Roles and Key Skills to Monetize Machine Learning',
 'From Open-Source Maintainer to Founder']

A/ `The Essentials of Public Speaking for Career in Data Science`